lib/experimental/interactivity-api/directive-processing.php

```diff
@@ -103,3 +168,156 @@ function gutenberg_interactivity_evaluate_reference( $path, array $context = arr
+			$content .= $inner_content;
+		} else {
+			// This is an inner block. It may be an interactive block or a
+			// non-interactive block.
+			$content                   .= '<wp-inner-blocks-' . $block_index . '></wp-inner-blocks-' . $block_index . '>';
+			$interactive_inner_blocks[] = $interactive_block['innerBlocks'][ $block_index ];
+			$block_index               += 1;
+		}
+	}
+	return gutenberg_process_interactive_html( $content, $interactive_inner_blocks, $context, $directives );
+}
+
+/**
+ * Traverses the HTML of a non-interactive block,
+ * skipping the processing and returning the content.
+ * For the inner blocks, it calls the corresponding function
+ * depending on the wrapper type.
+ *
+ * @param array                $non_interactive_block The non-interactive block to process.
+ * @param WP_Directive_Context $context The context to use when processing.
+ * @param array                $directives The directives to apply.
+ *
+ * @return string The processed HTML.
+ */
+function gutenberg_process_non_interactive_block( $non_interactive_block, $context, $directives ) {
+	$block_index = 0;
+	$content     = '';
+	foreach ( $non_interactive_block['innerContent'] as $inner_content ) {
+		if ( is_string( $inner_content ) ) {
+			// This content belongs to a non interactive block and therefore it cannot
+			// contain directives. We add the HTML directly to the final output.
+			$content .= $inner_content;
+		} else {
+			// This is an inner block. It may be an interactive block or a
+			// non-interactive block.
+			$inner_block  = $non_interactive_block['innerBlocks'][ $block_index ];
+			$block_index += 1;
+
+			if ( 'core/interactivity-wrapper' === $inner_block['blockName'] ) {
+				$content .= gutenberg_process_interactive_block( $inner_block, $context, $directives );
+			} elseif ( 'core/non-interactivity-wrapper' === $inner_block['blockName'] ) {
+				$content .= gutenberg_process_non_interactive_block( $inner_block, $context, $directives );
+			}
+		}
+	}
+	return $content;
+}
+
+/**
+ * Processes interactive HTML by applying directives to the HTML tags.
+ *
+ * @param string $html The HTML to process.
+ * @param array  $inner_blocks The inner blocks to process.
+ * @param mixed  $context The context to use when processing.
+ * @param array  $directives The directives to apply.
+ *
+ *  This function processes an HTML string by applying directives to the HTML tags.
+ *  It uses the WP_Directive_Processor class to parse the HTML and apply the directives.
+ *  The directives are specified in the $directives array and are applied to the tags that
+ *  have the corresponding data attribute.
+ *  If a tag contains a 'WP-INNER-BLOCKS' string and there are inner blocks to process, the function
+ *  processes these inner blocks and replaces the 'WP-INNER-BLOCKS' tag in the HTML with those blocks.
+ *
+ * @return string The processed HTML.
+ */
+function gutenberg_process_interactive_html( $html, $inner_blocks, $context, $directives ) {
```

Comment:
Related to https://github.com/WordPress/gutenberg/issues/56803#issuecomment-1864971003, could we move `$inner_blocks` to be the last parameter of this function and make it optional? Here's why:
* It's the only parameter that the function can work without (see `! empty` check in the function code).
* It would make it intuitive to use the function outside of a block context as well, to simply process an HTML string that contains interactivity attributes.

Reply:
Done!

@DAreRodz do you think that $context can be optional here too? If the developer does not provide it, we can declare it inside the function.

Something like:
```php
if ( ! isset( $context ) ) {
    $context = new WP_Directive_Context();
}
```